src/test/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcherTest.java

```diff
@@ -79,4 +92,349 @@ public void testShouldUpdate() throws IOException {
+    try (Repository repository =
+        FileRepositoryBuilder.create(project3Dir.resolve(".git").toFile())) {
+      repository.create();
+      createRepoTestFile(repository, thirdProject, "testFile3");
+
+      fetcher.addRepositoryInfoForTesting(thirdProject, project3Dir);
+
+      localRepositoryFor(thirdProject, 3);
+      assertEquals(fetcher.github(), github2);
+    }
+  }
+
+  @Test
+  public void testMultiThreadReadAndUpdate() throws IOException, InterruptedException {
+    ScheduledExecutorService executorService = Executors.newScheduledThreadPool(2);
+
+    GitHubProject firstProject = new GitHubProject("test", "project");
+    Path project1Dir = projectDir(firstProject.name());
+    try (Repository repository =
+        FileRepositoryBuilder.create(project1Dir.resolve(".git").toFile())) {
+      repository.create();
+
+      createRepoTestFile(repository, firstProject, "testFile");
+
+      fetcher.addRepositoryInfoForTesting(firstProject, project1Dir);
+
+      CountDownLatch latch = new CountDownLatch(1);
+
+      Runnable cleanTask = () -> {
+        try {
+          fetcher.cleanup((url, repo, total) -> {
+            latch.countDown();
+            return true;
+          });
+          checkCleanUp(firstProject, 0);
+        } catch (IOException e) {
+          throw new IllegalStateException("Clean up exception");
+        }
+      };
+
+      GitHubProject secondProject = new GitHubProject("test", "project2");
+
+      Runnable loadRepoTask = () -> {
+        try {
+          assertTrue(latch.await(10, TimeUnit.SECONDS));
+          fetcher.localRepositoryFor(secondProject);
+          checkLocalRepository(secondProject, 1);
+        } catch (InterruptedException | IOException e) {
+          throw new IllegalStateException("Load repository exception");
+        }
+      };
+
+      executorService.execute(cleanTask);
+      executorService.execute(loadRepoTask);
+    } finally {
+      executorService.shutdown();
+    }
+  }
+
+  @Test
+  public void testMultiThreadLocalRepoInfoTest() throws IOException, InterruptedException {
+    ExecutorService executorService = Executors.newFixedThreadPool(10);
+
+    GitHubProject project = new GitHubProject("test", "project");
+    Path projectDir = projectDir(project.name());
+    try (
+        Repository repository = FileRepositoryBuilder.create(projectDir.resolve(".git").toFile())) {
+      repository.create();
+
+      createRepoTestFile(repository, project, "testFile");
+
+      fetcher.addRepositoryInfoForTesting(project, projectDir);
+
+      int numberOfThreads = 25;
+      CountDownLatch latch = new CountDownLatch(numberOfThreads);
+      for (int i = 0; i < numberOfThreads; i++) {
+        executorService.submit(() -> {
+          try {
+            localRepositoryFor(project);
+          } catch (IOException e) {
+            throw new IllegalStateException("Load repository exception");
+          }
+          latch.countDown();
+        });
+      }
+      latch.await();
```

Comment:
I think the test can just use `ExecutinService.awaitTermination()`, can't it?

Reply:
`ExecutionService.awaitTermination()` takes in `timeout` argument. Which would be hard termination.
I wanted to make sure that all the threads are executed at least once.

Reply:
Using this check to confirm if all the threads have completed the task with `assertTrue(latch.await(5, TimeUnit.SECONDS));`

Reply:
If the threads didn't finish, then `await()` return false, and `assertTrue()` throws an exception.

If the threads didn't finish, then `ExecutinService.awaitTermination()` would throw an exception.

In both cases, the tests should fail.